src/firmware_new/tools/communication/transfer_media/transfer_medium.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
+#include "software/multithreading/thread_safe_buffer.h"
+
+/*
+ * A TransferMedium is an abstraction around the medium used to communicate
+ * between computers and robots. Any medium such as UART, SPI, Ethernet/WiFi
+ * used to transfer data to and from robots can extend this class.
+ *
+ */
+class TransferMedium
+{
+   public:
+    TransferMedium()          = default;
+    virtual ~TransferMedium() = default;
+
+    /*
+     * Packages the data and sends it through the medium
+     *
+     * @param data The data to be sent over the medium
+     */
+    virtual void send_data(const std::string& data)
+    {
+        throw std::logic_error("base TransferMedium send_data called");
+    }
+
+    /*
+     * Calls the provided callback everytime a new packet is received
+     *
+     * NOTE: this function does NOT block, it returns immediately after
+     * registering the callback
+     */
+    virtual void receive_data_async(std::function<void(std::string)> receive_callback)
+    {
+        throw std::logic_error("base TransferMedium receive_data_async called");
+    }
```

Comment:
if you're just trying to enforce that these functions be overriden, should be able to just add `= 0` after them and then it should be compile-time enforced I think?

Reply:
I tried this initially, but I wanted to be able to have a generic `TransferMedium` as a private member inside the communicator class, wasn't letting me do that when the functions where pure virtual

Reply:
Yeah, generally the approach here is to store the `TransferMedium` as a pointer (preferable a `std::unique_ptr`). Could we do that here to be consistent with the rest of the repo? That also gives us the bonus of compile-time guarantees.

Reply:
I switched over to `unique_ptrs` earlier, but didn't realize this was possible

switched to pure virtual, works well